### ISSUE TITLE
[BUG] listenForDeviceOrientationChanges reports the wrong landscape orientation

### ISSUE THREAD

**Issue by RuudBurger** [gladiuscode/react-native-orientation-director#39]
We need to implement the following code to allow for both landscape orientation. Seems to be wrong on both Android and iOS.

```ts
RNOrientationDirector.lockTo(Orientation.landscapeRight);

RNOrientationDirector.listenForDeviceOrientationChanges(
  ({ orientation }) => {
    // The library reports the opoosite orientation, so we need to swap them...
    if (orientation === Orientation.landscapeLeft) {
      RNOrientationDirector.lockTo(Orientation.landscapeRight);
    } else if (orientation === Orientation.landscapeRight) {
      RNOrientationDirector.lockTo(Orientation.landscapeLeft);
    }
  },
);
```

**Comment by gladiuscode** [comment]
Hi @RuudBurger, 
to be honest this kind of behaviour has always surprised me too.

First things first: **Android is aligned to iOS, so I can confirm that it does behave the same**.

That said, the reason behind this behaviour can be found in Apple's documentation, specifically: 

- [UIInterfaceOrientation](https://developer.apple.com/documentation/uikit/uiinterfaceorientation#overview)
- [UIDeviceOrientation](https://developer.apple.com/documentation/uikit/uideviceorientation)

In UIInterfaceOrientation documentation, we can find the following warning:

![Screenshot 2024-11-05 at 19 33 11](https://github.com/user-attachments/assets/894ca106-4025-4fe8-ab9e-6a54c19705b2)

So I've deduced that when your device camera in pointing left, your device is on **landscapeLeft** while your interface orientation is on **landscapeRight** and viceversa, this is because your UI content is rotating to the opposite direction. (🤯 *this is me, learning about this at 3 am while developing this library*)

Surfing the internet, we can see that we are not alone, unfortunately we can find many discussions, such as the following one: [screen orientation - ios-landscape-left-vs-right-vs-wrong](https://stackoverflow.com/questions/55149745/ios-landscape-left-vs-right-vs-wrong).

I've decided to follow Apple's documentation and align the code such as it does returns opposite values between device landscape orientation and interface landscape orientation, matching native expectations.

Lastly, I believe that your code is indeed correct, you are listening for **device** orientation changes, meaning that you should lock your interface orientation to the opposite landscape value.

Do you agree? 
Let me know what you think please, I've looked into other libraries as well and it does seem to behave the same. 
I have yet to test them out, I could do it in the weekend and let you know.

Cheers!

**Comment by RuudBurger** [comment]
Thanks for the full explanation.  

I think the difference with UIInterfaceOrientation and UIDeviceOrientation is that this library only has 1 orientation enum, and not 2. So on Apple you could do `UIInterfaceOrientation.landscapeLeft === UIDeviceOrientation.landscapeRight` to compare them (not actual code).

This library only uses `Orientation` so I think the solution would be to switch it in `listenForDeviceOrientationChanges` by default OR have a separate enum (which I don't think is a good solution and would only complicate things).

I would expect that the below code would work.

```
RNOrientationDirector.listenForDeviceOrientationChanges(
  ({ orientation }) => {
   RNOrientationDirector.lockTo(orientation)
  },
);
```

It's a minor thing, so the code on our end works either way, just wanted to contribute back with this "bug". No worries if you want to keep it like it is.

**Comment by RuudBurger** [comment]
BTW, what we are actually solving is having a `RNOrientationDirector.lockTo(Orientation.landscape)` option, which, depending how you hold the device would just rotate it either `landscapeLeft` or `landscapeRight`.

So if we could have that, would also be nice.

**Comment by gladiuscode** [comment]
I did some tests, because I wanted to be sure about it and **I confirm that they are indeed
the opposite at runtime.**

To check them out, I've added the following logs inside the `onOrientationChanged` method:

```swift
  private func onOrientationChanged(uiDeviceOrientation: UIDeviceOrientation) {
    let deviceOrientation = utils.convertToOrientationFrom(deviceOrientation: uiDeviceOrientation)
    self.eventManager.sendDeviceOrientationDidChange(orientationValue: deviceOrientation.rawValue)
    lastDeviceOrientation = deviceOrientation

    print("orientation: \(uiDeviceOrientation)")
    print("orientation: \(utils.getInterfaceOrientation())")

    adaptInterfaceTo(deviceOrientation: deviceOrientation)
  }
```

And they report the following when the device is in Landscape Left:

```txt
orientation: UIDeviceOrientation(rawValue: 3)
orientation: UIInterfaceOrientation(rawValue: 3)
```

Looking inside their definition:

- UIDeviceOrientation(rawValue: 3) matches `case landscapeLeft = 3 // Device oriented horizontally, home button on the right`;
- UIInterfaceOrientation(rawValue: 3), instead, matches: `case landscapeRight = 3`;

So Apple is 100% inverting their value for our lovely sake.

This and your feedback brings me to the following reasoning:

1. I should not manually compute the interface to the device orientation. Instead I should directly use
   the native api through utils.getInterfaceOrientation and be done with it. Of course this needs some
   testing, but I think we could benefit from it, having less error prone code that mimics the native
   behaviour; (Probably on Android too, I should look into it)
2. Having to switch on the js is a pain, I can recognise that, more so since others libs seems to
   manually invert the values so that on the js side device orientation === interface orientation.
   **I am not a fan of this way**, because **in my opinion a library should simply wrap the native side
   and expose the platform APIs as they are, pro and cons.** But nothing prevents us from implementing
   a js public API that ease this kind of operation for the sake of the developer mind (mine too).
   That `if` is indeed kind of strange, I would argue that a React Native developer, not knowing
   about this native weird stuff, would be thrown off by that, like: "why on earth?";

So the next steps should be to:

1. Remove, if possible, that adaptInterfaceTo method and leverage the native api fully;
2. Implement a way so that `lockTo` can be customised;
3. Look into the `lockTo(landscape)` and how we could implement it;

For this last feature, I'm thinking about a second parameter to the lockTo method that allows the
developer to tell it if the orientation passed in is 'device' or 'interface', so that it can act
accordingly and invert them out.

Although I'm not sure if it makes it easier to understand and what should the default option value
be: 'device' or 'interface'?

What do you think about it?

It could also be a good idea to add a Troubleshooting or FAQ about this, so that people can refer
to it, to better understand it.
On this note, I've also done some tests on Android and there it does behaves the same as Apple...

Again, thank you @RuudBurger, I really appreciate any feedback that can improve the library.

**Comment by gladiuscode** [comment]
Hi @RuudBurger,

I've implemented all required changes and from version 2.0.1 you should be able to achieve the desired behaviour through the following implementation:

```ts
useEffect(() => {
    const listener = RNOrientationDirector.listenForDeviceOrientationChanges(
      ({ orientation }) => {
        if (RNOrientationDirector.isLockableOrientation(orientation)) {
          RNOrientationDirector.lockTo(orientation, OrientationType.device);
        }
      }
    );

    return () => {
      listener.remove();
    };
}, []);
```

It should now fix the issue.
